Revert "[FIX] account: Error when manually adding analytic account in the generated tax lines on an invoice"

alhashash opened the pull request:
This reverts commit ab797f0cd8f76be232fcd918eafa52b847102504.

OCB has already merged a fix for the issue with commit 7b7ca96e654b67481f824e2ccead757ab4ae81c6

The reverted fix may introduce a problem if multiple taxes have similar code but different analytic account or multiple invoice lines with tax that has no account.

This PR is related to #82

alhashash committed:
Revert "[FIX] account: Error when manually adding analytic account in
the generated tax lines on an invoice"

This reverts commit ab797f0cd8f76be232fcd918eafa52b847102504.

OCB has already merged a fix for the issue with commit
7b7ca96e654b67481f824e2ccead757ab4ae81c6

The reverted fix may introduce a problem if multiple taxes have 
similar code but different analytic account.

sbidoul commented:
@alhashash with this PR you are diverging from upstream.

Removing the analytic account from the key was done on purpose upstream and it was discussed with odoo R&D as explained in ab797f0 and related launchpad issue. 

Can you elaborate on the use case causing a problem for you?

alhashash commented:
@sbidoul This conflicts with 18b7b28bc1ee676a8d245f50821d4a9aeb844158 (authored by you). I can see two problematic cases:
- Multiple lines with different analytic accounts and a tax with no account as per the related launchpad issue: One tax line is generated with the analytic account of the first line. 
- Multiple taxes with similar or no tax/base codes but different analytic accounts: One tax line is generated using the analytic account of tax that came first. 

The original issue was using the invoice line analytic account for non-refundable taxes. Your commit fixed the problem but ab797f0cd8f76be232fcd918eafa52b847102504 made it worse and introduced a new problem. 

I think the complete fix is to make the field read-only (as it is now in OCB) and allow editing it in the view for lines with `manual==True` using `attrs`.

alhashash committed:
Prevent changing analytic account in non-manual invoice tax lines


Analytic account should be edited only if the line is manually typed.

this is related to issue #374 and
https://bugs.launchpad.net/ocb-addons/+bug/1084822

alhashash commented:
I added another commit to remove the field readonly attribute (to match the upstream) and control the edit-ability in the view.

sbidoul commented:
@alhashash I see your point.

However not allowing to change the analytic account on taxes where manual=False will not work either.
In a situation where you have an invoice line with expense account A and a tax going partly to expense account B, the user must have the possibility to set the analytic account manually on expense account B too.

A possibility could be to make the check_tax_lines method smarter and ignoring the analytic account in that method, ie when checking only.

alhashash commented:
@sbidoul I agree. Smarter `check_tax_lines()` my be a better solution. But changing analytic account on tax was never allowed and commit 7b7ca96e654b67481f824e2ccead757ab4ae81c6 breaks current behavior and the business case it was supposed to solve. 

I think the common cases are:
1. Use the invoice line analytic account, that was the actual goal of [the Launchpad issue](https://launchpad.net/bugs/1084822).
2. Use a tax specific analytic account.  

Case 2 is already supported in tax definition. Case 1 is partially supported by your commit except when the tax has a financial account defined like the example you described. 

I think a better solution for case 1 is to a add a check box to tax definition to explicitly specify that this tax takes the analytic account of the invoice line. An update script can be included to mark this field as True for all taxes that have no analytic account to keep the current behavior unchanged. 

BTW, I made an [upstream PR](https://github.com/odoo/odoo/pull/3523) based on this PR.

gurneyalex commented:
@alhashash could you rebase so that we can access a runbot build?

alhashash commented:
@gurneyalex rebase is done

pedrobaeza commented:
This is not needed as the conflict was solved directly, so I close.
